Compiler warnings

I'm getting the following compiler warnings:

```bash
In file included from /data/prog/forks/xtensor/test/test_xdynamic_view.cpp:16:
In file included from /data/prog/forks/xtensor/include/xtensor/xrandom.hpp:23:
/data/prog/forks/xtensor/include/xtensor/xview.hpp:1688:35: warning: implicit conversion changes signedness: 'unsigned long' to 'typename
      iterator_traits<const long *>::difference_type' (aka 'long') [-Wsign-conversion]
            std::advance(first, k - xt::integral_count_before<S...>(i));
            ~~~                 ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/data/prog/forks/xtensor/include/xtensor/xview.hpp:1090:22: note: in instantiation of function template specialization
      'xt::xview<xt::xarray_container<xt::uvector<float, std::__1::allocator<float> >, xt::layout_type::row_major, xt::svector<unsigned
      long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, xt::xkeep_slice<long>, xt::xall<unsigned long>
      >::make_index<const long *>' requested here
        auto index = make_index(first, last);
                     ^
/data/prog/forks/xtensor/include/xtensor/xstrided_view_base.hpp:884:25: note: in instantiation of function template specialization
      'xt::xview<xt::xarray_container<xt::uvector<float, std::__1::allocator<float> >, xt::layout_type::row_major, xt::svector<unsigned
      long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, xt::xkeep_slice<long>, xt::xall<unsigned long>
      >::element<const long *>' requested here
            return m_e->element(m_index.cbegin(), m_index.cend());
                        ^
/data/prog/forks/xtensor/include/xtensor/xdynamic_view.hpp:533:16: note: in instantiation of member function
      'xt::detail::flat_expression_adaptor<xt::xview<xt::xarray_container<xt::uvector<float, std::__1::allocator<float> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &,
      xt::xkeep_slice<long>, xt::xall<unsigned long> > >::operator[]' requested here
        return base_type::storage()[static_cast<size_type>(offset)];
               ^
/data/prog/forks/xtensor/include/xtensor/xiterator.hpp:881:21: note: in instantiation of function template specialization
      'xt::xdynamic_view<xt::xview<xt::xarray_container<xt::uvector<float, std::__1::allocator<float> >, xt::layout_type::row_major,
      xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, xt::xkeep_slice<long>,
      xt::xall<unsigned long> > &, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::layout_type::dynamic,
      xt::detail::flat_expression_adaptor<xt::xview<xt::xarray_container<xt::uvector<float, std::__1::allocator<float> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &,
      xt::xkeep_slice<long>, xt::xall<unsigned long> > > >::element<const unsigned long *>' requested here
        return p_e->element(m_index.cbegin(), m_index.cend());
                    ^
/data/prog/forks/xtensor/include/xtensor/xiterator.hpp:1051:16: note: in instantiation of member function 'xt::xindexed_stepper<const
      xt::xdynamic_view<xt::xview<xt::xarray_container<xt::uvector<float, std::__1::allocator<float> >, xt::layout_type::row_major,
      xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, xt::xkeep_slice<long>,
      xt::xall<unsigned long> > &, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::layout_type::dynamic,
      xt::detail::flat_expression_adaptor<xt::xview<xt::xarray_container<xt::uvector<float, std::__1::allocator<float> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &,
      xt::xkeep_slice<long>, xt::xall<unsigned long> > > >, true>::operator*' requested here
        return *m_st;
               ^
/data/prog/forks/xtensor/include/xtensor/xassign.hpp:576:44: note: (skipping 4 contexts in backtrace; use -ftemplate-backtrace-limit=0 to
      see all)
            *dst = static_cast<value_type>(*src);
                                           ^
/data/prog/forks/xtensor/include/xtensor/xassign.hpp:201:40: note: in instantiation of function template specialization
      'xt::xexpression_assigner<xt::xtensor_expression_tag>::assign_xexpression<xt::xexpression<xt::xarray_container<xt::uvector<float,
      std::__1::allocator<float> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> >, xt::xexpression<xt::xdynamic_view<xt::xview<xt::xarray_container<xt::uvector<float,
      std::__1::allocator<float> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &, xt::xkeep_slice<long>, xt::xall<unsigned long> > &, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::layout_type::dynamic,
      xt::detail::flat_expression_adaptor<xt::xview<xt::xarray_container<xt::uvector<float, std::__1::allocator<float> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &,
      xt::xkeep_slice<long>, xt::xall<unsigned long> > > > > >' requested here
            xexpression_assigner<tag>::assign_xexpression(e1, e2);
                                       ^
/data/prog/forks/xtensor/include/xtensor/xsemantic.hpp:587:13: note: in instantiation of function template specialization
      'xt::assign_xexpression<xt::xarray_container<xt::uvector<float, std::__1::allocator<float> >, xt::layout_type::row_major,
      xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag>,
      xt::xdynamic_view<xt::xview<xt::xarray_container<xt::uvector<float, std::__1::allocator<float> >, xt::layout_type::row_major,
      xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, xt::xkeep_slice<long>,
      xt::xall<unsigned long> > &, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::layout_type::dynamic,
      xt::detail::flat_expression_adaptor<xt::xview<xt::xarray_container<xt::uvector<float, std::__1::allocator<float> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &,
      xt::xkeep_slice<long>, xt::xall<unsigned long> > > > >' requested here
        xt::assign_xexpression(*this, e);
            ^
/data/prog/forks/xtensor/include/xtensor/xsemantic.hpp:453:37: note: in instantiation of function template specialization
      'xt::xcontainer_semantic<xt::xarray_container<xt::uvector<float, std::__1::allocator<float> >, xt::layout_type::row_major,
      xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag>
      >::assign_xexpression<xt::xdynamic_view<xt::xview<xt::xarray_container<xt::uvector<float, std::__1::allocator<float> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &,
      xt::xkeep_slice<long>, xt::xall<unsigned long> > &, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::layout_type::dynamic, xt::detail::flat_expression_adaptor<xt::xview<xt::xarray_container<xt::uvector<float,
      std::__1::allocator<float> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &, xt::xkeep_slice<long>, xt::xall<unsigned long> > > > >' requested here
        return this->derived_cast().assign_xexpression(e);
                                    ^
/data/prog/forks/xtensor/include/xtensor/xarray.hpp:457:24: note: in instantiation of function template specialization
      'xt::xsemantic_base<xt::xarray_container<xt::uvector<float, std::__1::allocator<float> >, xt::layout_type::row_major,
      xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag>
      >::assign<xt::xdynamic_view<xt::xview<xt::xarray_container<xt::uvector<float, std::__1::allocator<float> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &,
      xt::xkeep_slice<long>, xt::xall<unsigned long> > &, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::layout_type::dynamic, xt::detail::flat_expression_adaptor<xt::xview<xt::xarray_container<xt::uvector<float,
      std::__1::allocator<float> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &, xt::xkeep_slice<long>, xt::xall<unsigned long> > > > >' requested here
        semantic_base::assign(e);
                       ^
/data/prog/forks/xtensor/test/test_xdynamic_view.cpp:260:21: note: in instantiation of function template specialization
      'xt::xarray_container<xt::uvector<float, std::__1::allocator<float> >, xt::layout_type::row_major, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag>::xarray_container<xt::xdynamic_view<xt::xview<xt::xarray_container<xt::uvector<float,
      std::__1::allocator<float> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>,
      xt::xtensor_expression_tag> &, xt::xkeep_slice<long>, xt::xall<unsigned long> > &, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::layout_type::dynamic,
      xt::detail::flat_expression_adaptor<xt::xview<xt::xarray_container<xt::uvector<float, std::__1::allocator<float> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &,
      xt::xkeep_slice<long>, xt::xall<unsigned long> > > > >' requested here
        auto res  = xt::xarray<float>(xt::dynamic_view(v, {xt::keep(0)}));
                    ^
1 warning generated.
```

---------------

```bash
/data/prog/forks/xtensor/test/test_xnpy.cpp:66:37: warning: 'tmpnam' is deprecated: This function is provided for compatibility reasons
      only. Due to security concerns inherent in the design of tmpnam(3), it is highly recommended that you use mkstemp(3) instead.
      [-Wdeprecated-declarations]
        std::string filename = std::tmpnam(nullptr);
                                    ^
/usr/include/stdio.h:186:1: note: 'tmpnam' has been explicitly marked deprecated here
__deprecated_msg("This function is provided for compatibility reasons only.  Due to security concerns inherent in the design of tmpn...
^
/usr/include/sys/cdefs.h:180:48: note: expanded from macro '__deprecated_msg'
        #define __deprecated_msg(_msg) __attribute__((deprecated(_msg)))
                                                      ^
1 warning generated.
```

---------------

```bash
/data/prog/forks/xtensor/test/test_xrandom.cpp:12:32: warning: unknown warning group '-Wmaybe-uninitialized', ignored
      [-Wunknown-warning-option]
#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
                               ^
1 warning generated.
```

----------

```bash
/data/prog/clone/xtensor/include/xtensor/xoperation.hpp:631:29: warning: comparison of integers of different signs: 'const long' and 'const unsigned long' [-Wsign-compare]
      'const unsigned long' [-Wsign-compare]
            res = (*iter1++ == *iter2++);
                   ~~~~~~~~ ^  ~~~~~~~~
/usr/local/include/gtest/gtest.h:1392:11: note: in instantiation of function template specialization
      'xt::operator==<xt::xview<xt::xarray_container<xt::uvector<long, std::__1::allocator<long> >, xt::layout_type::row_major,
      xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, int, xt::xall<unsigned long>,
      xt::xall<unsigned long> >, xt::xview<xt::xarray_container<xt::uvector<unsigned long, std::__1::allocator<unsigned long> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &,
      int, xt::xall<unsigned long>, xt::xall<unsigned long> > >' requested here
  if (lhs == rhs) {
          ^
/usr/local/include/gtest/gtest.h:1421:12: note: in instantiation of function template specialization
      'testing::internal::CmpHelperEQ<xt::xview<xt::xarray_container<xt::uvector<long, std::__1::allocator<long> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &,
      int, xt::xall<unsigned long>, xt::xall<unsigned long> >, xt::xview<xt::xarray_container<xt::uvector<unsigned long,
      std::__1::allocator<unsigned long> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>,
      true>, xt::xtensor_expression_tag> &, int, xt::xall<unsigned long>, xt::xall<unsigned long> > >' requested here
    return CmpHelperEQ(lhs_expression, rhs_expression, lhs, rhs);
           ^
/data/prog/forks/xtensor/test/test_extended_xsort.cpp:485:9: note: in instantiation of function template specialization
      'testing::internal::EqHelper<false>::Compare<xt::xview<xt::xarray_container<xt::uvector<long, std::__1::allocator<long> >,
      xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &,
      int, xt::xall<unsigned long>, xt::xall<unsigned long> >, xt::xview<xt::xarray_container<xt::uvector<unsigned long,
      std::__1::allocator<unsigned long> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>,
      true>, xt::xtensor_expression_tag> &, int, xt::xall<unsigned long>, xt::xall<unsigned long> > >' requested here
        EXPECT_EQ(xt::view(py_a0, 2, all(), all()), xt::view(a0, 2, all(), all()));
        ^
/usr/local/include/gtest/gtest.h:1924:63: note: expanded from macro 'EXPECT_EQ'
                      EqHelper<GTEST_IS_NULL_LITERAL_(val1)>::Compare, \
                                                              ^
```

--- a/include/xtensor/xbuilder.hpp
+++ b/include/xtensor/xbuilder.hpp
@@ -214,7 +214,8 @@ namespace xt
             template <class It>
             inline T element(It first, It) const
             {
-                return m_start + m_step * T(*first);
+                // Avoids warning when T = char (because char + char => int!)
+                return static_cast<T>(m_start + m_step * T(*first));
             }
 
             template <class E>

--- a/include/xtensor/xgenerator.hpp
+++ b/include/xtensor/xgenerator.hpp
@@ -410,7 +410,8 @@ namespace xt
     {
         using vtype = typename shape_type::value_type;
         xt::dynamic_shape<vtype> sh(shape.size());
-        typename std::decay_t<O>::value_type accumulator(1);
+        using int_type = typename std::decay_t<O>::value_type;
+        int_type accumulator(1);
         std::size_t neg_idx = 0;
         std::size_t i = 0;
         for(std::size_t j = 0; j != shape.size(); ++j, ++i)
@@ -429,7 +430,7 @@ namespace xt
         }
         if(accumulator < 0)
         {
-            sh[neg_idx] = this->size() / static_cast<size_type>(std::abs(accumulator));
+            sh[neg_idx] = this->size() / static_cast<size_type>(std::make_unsigned_t<int_type>(std::abs(accumulator)));
         }
         return sh;
     }

--- a/include/xtensor/xview.hpp
+++ b/include/xtensor/xview.hpp
@@ -1498,6 +1498,7 @@ namespace xt
     inline auto xview<CT, S...>::make_index(It first, It last) const -> base_index_type
     {
         auto index = xtl::make_sequence<base_index_type>(m_e.dimension(), 0);
+        using diff_type = typename std::iterator_traits<It>::difference_type;
         using ivalue_type = typename base_index_type::value_type;
         auto func1 = [&first](const auto& s) noexcept {
             return get_slice_value(s, first);
@@ -1513,7 +1514,7 @@ namespace xt
 
             // need to advance captured `first`
             first = first_copy;
-            std::advance(first, k - xt::integral_count_before<S...>(i));
+            std::advance(first, static_cast<diff_type>(k - xt::integral_count_before<S...>(i)));
 
             if (first < last)
             {

--- a/test/test_extended_xsort.cpp
+++ b/test/test_extended_xsort.cpp
@@ -42,7 +42,7 @@ namespace xt
     template <class X, class Y, class Z>
     bool check_argpartition_equal(const X& data, const Y& a1, const Z& a2, std::size_t kth)
     {
-        auto p = a1[kth];
+        auto p = static_cast<std::size_t>(a1[kth]);
         EXPECT_EQ(p, std::size_t(a2[kth]));
         auto el = data[static_cast<std::size_t>(a1[kth])];
         for (std::size_t i = 0; i < kth; ++i)
@@ -104,7 +104,7 @@ namespace xt
         check_argpartition_equal(py_a, py_a19, xt::argpartition(py_a, 19), 19);
 
         // py_median = np.median(a)
-        double py_median = 300.0;
+        long py_median = 300;
         EXPECT_EQ(py_median, xt::median(py_a));
     }
 
@@ -482,9 +482,9 @@ namespace xt
         auto a1 = xt::argpartition(py_a, 4, 1);
         auto a2 = xt::argpartition(py_a, 3, 2);
 
-        EXPECT_EQ(xt::view(py_a0, 2, all(), all()), xt::view(a0, 2, all(), all()));
-        EXPECT_EQ(xt::view(py_a1, all(), 4, all()), xt::view(a1, all(), 4, all()));
-        EXPECT_EQ(xt::view(py_a2, all(), all(), 3), xt::view(a2, all(), all(), 3));
+        EXPECT_EQ(xt::cast<std::size_t>(xt::view(py_a0, 2, all(), all())), xt::view(a0, 2, all(), all()));
+        EXPECT_EQ(xt::cast<std::size_t>(xt::view(py_a1, all(), 4, all())), xt::view(a1, all(), 4, all()));
+        EXPECT_EQ(xt::cast<std::size_t>(xt::view(py_a2, all(), all(), 3)), xt::view(a2, all(), all(), 3));
     }
 
     /*py
@@ -721,13 +721,13 @@ namespace xt
         auto a1 = xt::argpartition(py_a, {1, 4}, 1);
         auto a2 = xt::argpartition(py_a, {1, 3}, 2);
 
-        EXPECT_EQ(xt::view(py_a0, 2, all(), all()), xt::view(a0, 2, all(), all()));
-        EXPECT_EQ(xt::view(py_a1, all(), 4, all()), xt::view(a1, all(), 4, all()));
-        EXPECT_EQ(xt::view(py_a2, all(), all(), 3), xt::view(a2, all(), all(), 3));
+        EXPECT_EQ(xt::cast<std::size_t>(xt::view(py_a0, 2, all(), all())), xt::view(a0, 2, all(), all()));
+        EXPECT_EQ(xt::cast<std::size_t>(xt::view(py_a1, all(), 4, all())), xt::view(a1, all(), 4, all()));
+        EXPECT_EQ(xt::cast<std::size_t>(xt::view(py_a2, all(), all(), 3)), xt::view(a2, all(), all(), 3));
 
-        EXPECT_EQ(xt::view(py_a0, 1, all(), all()), xt::view(a0, 1, all(), all()));
-        EXPECT_EQ(xt::view(py_a1, all(), 1, all()), xt::view(a1, all(), 1, all()));
-        EXPECT_EQ(xt::view(py_a2, all(), all(), 1), xt::view(a2, all(), all(), 1));
+        EXPECT_EQ(xt::cast<std::size_t>(xt::view(py_a0, 1, all(), all())), xt::view(a0, 1, all(), all()));
+        EXPECT_EQ(xt::cast<std::size_t>(xt::view(py_a1, all(), 1, all())), xt::view(a1, all(), 1, all()));
+        EXPECT_EQ(xt::cast<std::size_t>(xt::view(py_a2, all(), all(), 1)), xt::view(a2, all(), all(), 1));
     }
 }
 

--- a/test/test_xnpy.cpp
+++ b/test/test_xnpy.cpp
@@ -61,16 +61,15 @@ namespace xt
             fn1_contents.size() == fn2_contents.size();
     }
 
-    std::string get_filename()
+    std::string get_filename(int n)
     {
-        std::string filename = std::tmpnam(nullptr);
-        filename += ".npy";
+        std::string filename = "files/xnpy_files/test_dump_" + std::to_string(n) + ".npy";
         return filename;
     }
 
     TEST(xnpy, dump)
     {
-        std::string filename = get_filename();
+        std::string filename = get_filename(0);
         xarray<bool> barr = {{{0, 0, 1},
                               {1, 1, 0},
                               {1, 0, 1}},
@@ -93,7 +92,7 @@ namespace xt
         EXPECT_TRUE(compare_binary_files(filename, compare_name));
         std::remove(filename.c_str());
 
-        filename = get_filename();
+        filename = get_filename(1);
         dump_npy(filename, ularr);
         auto ularrcpy = load_npy<uint64_t>(filename);
         EXPECT_TRUE(all(equal(ularr, ularrcpy)));

--- a/test/test_xrandom.cpp
+++ b/test/test_xrandom.cpp
@@ -7,7 +7,7 @@
 ****************************************************************************/
 
 #include "gtest/gtest.h"
-#ifdef __GNUC__
+#if (defined(__GNUC__) && !defined(__clang__))
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
 #include "xtensor/xrandom.hpp"

--- a/test/test_xreducer.cpp
+++ b/test/test_xreducer.cpp
@@ -16,7 +16,14 @@
 #include "xtensor/xreducer.hpp"
 #include "xtensor/xview.hpp"
 #include "xtensor/xmanipulation.hpp"
+#if (defined(__GNUC__) && !defined(__clang__))
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
 #include "xtensor/xrandom.hpp"
+#pragma GCC diagnostic pop
+#else
+#include "xtensor/xrandom.hpp"
+#endif
 
 namespace xt
 {
@@ -119,7 +126,7 @@ namespace xt
         const xreducer_features::shape_type& s = features.m_red.shape();
         std::size_t nb_iter = 1;
         nb_iter = std::accumulate(s.cbegin(), s.cend(), nb_iter, std::multiplies<std::size_t>());
-        std::advance(iter, nb_iter);
+        std::advance(iter, static_cast<std::ptrdiff_t>(nb_iter));
         EXPECT_EQ(iter_end, iter);
     }
 

--- a/test/test_xsort.cpp
+++ b/test/test_xsort.cpp
@@ -290,7 +290,7 @@ namespace xt
         auto ma0 = median(a, 0);
         auto ma1 = median(a, 1);
 
-        EXPECT_EQ(mall, 5);
+        EXPECT_EQ(mall, 5.f);
 
         xt::xtensor<float, 1> ma0_exp = {5, 5, 5, 5};
         xt::xtensor<float, 1> ma1_exp = {2.5, 1.5, 9., 12., 5.};

